### ISSUE TITLE
Gradient function not found for tf.topk

### ISSUE THREAD

**Issue by kwonoh** [tensorflow/tfjs#1601]
To get help from the community, we encourage using Stack Overflow and the [`tensorflow.js`](https://stackoverflow.com/questions/tagged/tensorflow.js) tag.

#### TensorFlow.js version
1.1.2

#### Browser version
All

#### Describe the problem or feature request
Error: Cannot compute gradient: gradient function not found for topk.

#### Code to reproduce the bug / link to feature request
Gradient function is not defined in [topk.ts](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/topk.ts) nor [topk_impl.ts](https://github.com/tensorflow/tfjs-core/blob/master/src/backends/topk_impl.ts).

```js
const f = x => tf.topk(x).values;
const g = tf.grad(f);
const x = tf.tensor1d([2, 3]);
g(x).print(); // -> Cannot compute gradient: gradient function not found for topk.
```

**Comment by kwonoh** [comment]
Gradient of topk in tensorflow: https://github.com/tensorflow/tensorflow/blob/e79459acdbcd97c4adc888ee7452c19f38178ee0/tensorflow/python/ops/nn_grad.py#L1069-L1112

**Comment by kwonoh** [comment]
Here is my topk function with gradients using `tf.customGrad`.

```python
const topkGrad = (dy, saved) => {
    const [x, indices] = saved;

    const inShape = x.shape;
    const indShape = indices.shape;
    const inLastDim = x.shape[x.shape.length - 1];
    const indLastDim = indShape[indShape.length - 1];

    const ind2d = indices.reshape([-1, indLastDim]);
    const outerDim = ind2d.shape[0];

    const ind = ind2d
                    .add(tf.range(0, outerDim * inLastDim, inLastDim, 'int32')
                             .expandDims(-1))
                    .reshape([-1]);

    return tf
        .scatterND(
            ind.expandDims(-1), dy.reshape([-1]),
            [tf.prod(inShape).arraySync()])
        .reshape(inShape);
};


const topk = tf.customGrad((x, k, save) => {
    k = k.arraySync();  // k needs to be a number
    const {values, indices} = tf.topk(x, k, true);
    save([x, indices]);
    return {value: values, gradFunc: topkGrad};
});


const sort = tf.customGrad((x, save) => {
    const k = x.shape[x.shape.length - 1];
    const {values, indices} = tf.topk(x, k, true);
    save([x, indices]);
    return {value: values, gradFunc: topkGrad};
});


x = tf.tensor2d([1, 2, 3, 4], [2, 2]);
k = tf.scalar(2);  // The args passed in customGrad must all be tensors
topk(x, k).print();

// I am not sure how to pass multiple arguments using tf.grad (not tf.grads)
// so, here is an example of a sort function
const dx = tf.grad(x => sort(x));
dx(x).print();
```

**Comment by DirkToewe** [comment]
If backpropagation is implemented for tf.gather, it might be simpler and more concise to re-implement tf.topk in terms of it.
```js
function topk(/*tf.Tensor*/x, /*int*/k)
{
  const shape = x.shape.slice(0,-1);

  const [n] = x.shape.slice(  -1),
         m  =   shape.reduce( (k,l) => k*l, 1 );

  const data = x.dataSync(),
      sorted = new Int32Array(n),
     indices = new Int32Array(m*k);

  for( let i=0; i < m; ++i )
  {
    for( let j=0; j < n; j++ )
      sorted[j] = j + i*n;

    sorted.sort( (i,j) => data[j] - data[i] );

    for( let j=0; j < k; j++ )
      indices[j + i*k] = sorted[j];
  }

  shape.push(k);

  return x.flatten()
          .gather(indices)
          .reshape(shape);
}
```

**Comment by rthadur** [comment]
I was not able to reproduce the error [here](https://js.tensorflow.org/api/0.6.1/#grad) , please check and let us know of this is still an issue ? or else will be closing this issue soon, thank you

**Comment by kwonoh** [comment]
It's still an issue. 
See [this screenshot](https://www.dropbox.com/s/7jwtpli5a90aezu/Screenshot%202019-10-05%2019.10.31.png?dl=0).
I tested on [this page (tfjs 1.2.7)](https://js.tensorflow.org/api/1.2.7/#topk).

**Comment by rthadur** [comment]
Changes are implemented here https://github.com/tensorflow/tfjs/pull/6216 recently. Closing this request. Thank you
